### DISCUSSION TITLE
Does `Dispatcher.Post` guarantee the call order?

### DISCUSSION THREAD

**Discussion by ktxiaok** [AvaloniaUI/Avalonia#17821] [Q&A]
Let's say:
```csharp
Dispatcher.UIThread.Post(() => Console.WriteLine("1"));
Dispatcher.UIThread.Post(() => Console.WriteLine("2"));
Dispatcher.UIThread.Post(() => Console.WriteLine("3"));
```
Does the output have to be this:
```
1
2
3
```
Actually it is but I don't know if it's a definite behavior.

**Comment by thevortexcloud** [comment]
My understanding about how the whole dispatcher thing is it dumps everything into a queue. Which then is executed in order by a different thread. However because the queue is globally accessible from any thread there is nothing stopping something else from getting in between your calls. Also the dispatcher has a priority system which determines the order things are executed in. Higher priority things are executed before lower priority ones. You can specify the priority yourself using one of the overloads.

**Reply by ktxiaok** [reply]
Thanks for your comment! I'm talking about the same priority situation. The current implementation of the dispatcher may guarantee the call order but this behavior is undocumented. So if the implementation changes in the future then the call order may no longer be guaranteed. In short, I wonder if it is a definite behavior.

**Reply by thevortexcloud** [reply]
I _believe_ (although I don't really have any evidence of it) that the dispatcher is loosely based on WPFs. I also highly doubt the dispatcher will change significantly at any point in the future (I would also not be surprised if XPF requires the current behaviour). It would break almost everything if it suddenly was not queue based, which would require a very compelling reason to make the change.

But these are really questions for the core team.
